File: HackerRank/Interview/Arrays/MinimumSwaps2.cxx
#include <bits/stdc++.h>
// https://www.hackerrank.com/challenges/minimum-swaps-2/problem
// Tag(s): arrays, implementation
using namespace std;

#define watch(x) cout << (#x) << " is " << x << endl;

int getMinSwaps(vector<int> &arr);

int main(int argc, char const *argv[]) {
    int n;
    cin >> n;
    vector<int> arr(n+1);

    for (int i = 1; i <= n; ++i) cin >> arr[i];

    cout << getMinSwaps(arr) << endl;

    return 0;
}

int getMinSwaps(vector<int> &arr) {
    int n = arr.size()-1;
    int ans = 0;
    int indices[n+1];

    for (int i = 1; i <= n; ++i) indices[arr[i]] = i;

    for (int i = 1; i <= n; ++i) {
        if(arr[i] == i) continue;

        int index = indices[i];

        indices[i] = i;
        indices[arr[i]] = index;
        swap(arr[i], arr[index]);
        ++ans;
    }

    return ans;
}

File: HackerRank/Interview/Arrays/minimum_swaps_2.py
# https://www.hackerrank.com/challenges/minimum-swaps-2
# tags: 


def min_swaps(arr: list) -> int:
    """My solution
    Time: O(n)
    Space: O(n)
    """
    swaps = 0
    elem_pos_map = {element: pos for pos, element in enumerate(arr)}

    # print(elem_pos_map)

    for idx in range(len(arr)):
        if arr[idx] == (idx+1):
            continue
        # index of element to swap
        jdx = elem_pos_map[idx+1]

        # update map
        elem_pos_map[idx+1] = idx
        elem_pos_map[arr[idx]] = jdx

        arr[idx], arr[jdx] = arr[jdx], arr[idx]
        swaps += 1

    return swaps



if __name__ == "__main__":
    n = int(input())
    arr = list(map(int, input().split()))

    print(min_swaps(arr))
